bn-apps/cordapp-updates-distribution/corda-updates-core/src/main/kotlin/net/corda/businessnetworks/cordaupdates/core/CordappSyncer.kt

```diff
@@ -0,0 +1,122 @@
+     *              If not provided, then all missing versions of all CorDapps defined in the [syncerConf] will be downloaded.
+     *              If provided, then [syncerConf] is expected to contain a configured [CordappSource] for the CorDapp coordinates.
+     * @additionalConfigurationProperties additional parameters to be passed to [CordaMavenResolver]
+     * @return a list of CorDapp metadatas that have been resolved from local / remote repositories
+     *
+     * @throws [CordappSourceNotFoundException] if CorDapp source has not been found
+     */
+    fun syncCordapps(coordinatesWithoutVersion : String? = null, additionalConfigurationProperties : Map<String, Any> = mapOf()) : List<ArtifactMetadata> {
+        return if (coordinatesWithoutVersion == null) {
+            syncerConf.cordappSources.flatMap { syncTask ->
+                val resolver = CordaMavenResolver.create(syncerConf, syncTask, repositoryListener = repositoryListener, transferListener = transferListener)
+                syncTask.cordapps.map {
+                    resolver.downloadVersionRange("$it:[,)", additionalConfigurationProperties)
+                }
+            }
+        } else {
+            val source = syncerConf.findSourceFor(coordinatesWithoutVersion) ?: throw CordappSourceNotFoundException(coordinatesWithoutVersion)
+            val resolver = CordaMavenResolver.create(syncerConf, source)
+            listOf(resolver.downloadVersionRange(coordinatesWithoutVersion, additionalConfigurationProperties))
+        }
+    }
+
+    /**
+     * Returns a list of CorDapp versions available in the remote repository. syncerConf is expected to contain a configured [CordappSource] for the requested CorDapp.
+     *
+     * @param coordinatesWithRange full maven coordinates of a CorDapp with a version range, i.e. "net.corda:corda-finance:[0,2.5)".
+     * @param additionalConfigurationProperties additional parameters to be passed to [CordaMavenResolver]
+     */
+    fun getAvailableVersions(coordinatesWithRange : String, additionalConfigurationProperties : Map<String, Any> = mapOf()) : List<ArtifactMetadata> {
+        val syncTask = syncerConf.findSourceFor(coordinatesWithRange) ?: throw CordappSourceNotFoundException(coordinatesWithRange)
+        val resolver = CordaMavenResolver.create(syncerConf, syncTask)
+        return listOf(resolver.resolveVersionRange(coordinatesWithRange, additionalConfigurationProperties))
+    }
+}
+
+/**
+ * [CordappSource] allows, that allows to associate multiple CorDapps with a remote repository
+ */
+data class CordappSource(
+        val remoteRepoUrl : String,
+        val cordapps : List<String>,
+        val httpUsername : String? = null,
+        val httpPassword : String? = null)
+
+/**
+ * Configuration for [CordappSyncer]. Usually is red from settings.conf
+ */
+data class SyncerConfiguration(
+        val localRepoPath : String,
```

Comment:
How comes this object isn't passed into the CordappSyncer - you duplicate most of its content in the params to create

Reply:
You mean `SyncerConfiguration`? It is passed as a constructor parameter. 
```
class CordappSyncer(private val syncerConf : SyncerConfiguration,
                    private val repositoryListener : RepositoryListener? = null,
                    private val transferListener : TransferListener? = null) {
```